core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java

```diff
@@ -138,40 +140,49 @@ private Schema initializeOutputSchema() {
   /**
    * @param inputSchema Validates whether the keyfield, fieldname, and fieldvalue entered by the user is of type
    *                    String or Nullable String and present in the input schema or not.
+   * @param collector FailureCollector used to log all errors and return them to the user.
    * @return true
    */
-  private void validateInputFields(Schema inputSchema) {
+  private void validateInputFields(Schema inputSchema, FailureCollector collector) {
 
     if (inputSchema.getField(conf.getKeyField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Keyfield '%s' does not exist in input schema %s", conf.getKeyField(), inputSchema));
-    } else if (inputSchema.getField(conf.getNameField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Namefield '%s' does not exist in input schema %s", conf.getNameField(), inputSchema));
-    } else if (inputSchema.getField(conf.getValueField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Valuefield '%s' does not exist in input schema %s", conf.getValueField(), inputSchema));
+      collector.addFailure(String.format("Keyfield '%s' does not exist in input schema", conf.getKeyField()),
+              "Provide a field that is matches a field from the input schema").withConfigProperty("keyField");
+    } else {
+      Schema keyFieldSchema = inputSchema.getField(conf.getKeyField()).getSchema();
+
+      if (!((keyFieldSchema.isNullable() ? keyFieldSchema.getNonNullable().getType() : keyFieldSchema
+              .getType()).equals(Schema.Type.STRING))) {
+        collector.addFailure(String.format("Keyfield '%s' in the input record must be of type String or" +
```

Comment:
Key field '%s' in the input schema must be a string

Reply:
similar comments below for the other fields

Reply:
Done